pbs/usersync.go

```diff
@@ -47,30 +61,162 @@ func ParseUIDCookie(r *http.Request) *PBSCookie {
 		// corrupted cookie; we should reset
 		return &pc
 	}
-	if pc.OptOut || pc.UIDs == nil {
-		pc.UIDs = make(map[string]string) // empty map
+	if pc.optOut || pc.uids == nil {
+		pc.uids = make(map[string]string) // empty map
 	}
+
+	// Facebook sends us a sentinel value of 0 if the user isn't logged in.
+	// As a result, we've stored  "0" as the UID for many users in the audienceNetwork so far.
+	// Since users log in and out of facebook all the time, this will cause re-sync attempts until
+	// we get a non-zero value.
+	if pc.uids["audienceNetwork"] == "0" {
+		delete(pc.uids, "audienceNetwork")
+	}
+
 	return &pc
 }
 
-func SetUIDCookie(w http.ResponseWriter, pc *PBSCookie) {
-	j, _ := json.Marshal(pc)
+// NewPBSCookie returns an empty PBSCookie
+func NewPBSCookie() *PBSCookie {
+	return &PBSCookie{
+		uids:     make(map[string]string),
+		birthday: timestamp(),
+	}
+}
+
+// PBSCookie is the cookie used in Prebid Server.
+//
+// To get an instance of this from a request, use ParsePBSCookieFromRequest.
+// To write an instance onto a response, use SetCookieOnResponse.
+type PBSCookie struct {
```

Comment:
I only meant the core struct, thanks for moving. Agree that the general refactoring should be done in a separate PR.